File: council/providers/anthropic.py
import anthropic

from .base import Message, Provider


class AnthropicProvider(Provider):
    def __init__(self, model_id: str, api_key: str):
        super().__init__(model_id, api_key)
        self.client = anthropic.Anthropic(api_key=api_key)

    def send_message(self, messages: list[Message]) -> str:
        system_parts = []
        api_messages = []

        for msg in messages:
            if msg.role == "system":
                system_parts.append(msg.content)
            else:
                api_messages.append({"role": msg.role, "content": msg.content})

        response = self.client.messages.create(
            model=self.model_id,
            max_tokens=2048,
            system="\n\n".join(system_parts) if system_parts else anthropic.NOT_GIVEN,
            messages=api_messages,
        )
        return response.content[0].text

File: workers/src/providers/anthropic.ts
import Anthropic from '@anthropic-ai/sdk';
import type { Message } from '../types.ts';

/**
 * Streams a response from Anthropic Claude.
 * System messages are extracted from the messages array and passed via the
 * dedicated `system` parameter, which is the Anthropic-recommended approach.
 */
export async function* streamAnthropic(
  apiKey: string,
  modelId: string,
  messages: Message[],
): AsyncGenerator<string> {
  const client = new Anthropic({ apiKey });

  const systemParts = messages
    .filter((m) => m.role === 'system')
    .map((m) => m.content);
  const system = systemParts.length > 0 ? systemParts.join('\n\n') : undefined;

  const chatMessages = messages
    .filter((m) => m.role !== 'system')
    .map((m) => ({ role: m.role as 'user' | 'assistant', content: m.content }));

  const stream = client.messages.stream({
    model: modelId,
    max_tokens: 2048,
    ...(system ? { system } : {}),
    messages: chatMessages,
  });

  for await (const event of stream) {
    if (
      event.type === 'content_block_delta' &&
      event.delta.type === 'text_delta'
    ) {
      yield event.delta.text;
    }
  }
}
